internal/fs/buffer/inmemory_write_buffer.go

```diff
@@ -15,33 +15,165 @@
+	}
+	if b.current.buffer == nil {
+		b.current.buffer = make([]byte, 0, b.bufferSize)
+		b.current.offset.end = b.bufferSize
+	}
+	return nil
+}
+
+// Allocates memory to flushedBuffer if it hasn't been allocated already.
+func (b *InMemoryWriteBuffer) ensureFlushedBuffer() error {
+	if b.bufferSize == 0 {
+		return fmt.Errorf(ZeroSizeBufferError)
+	}
+	if b.flushed.buffer == nil {
+		b.flushed.buffer = make([]byte, 0, b.bufferSize)
+	}
+	return nil
 }
 
-func (b *InMemoryWriteBuffer) InitializeBuffer(sizeInMB int) {
-	ChunkSize = sizeInMB * MiB
-	if b.buffer == nil {
-		b.buffer = bytes.NewBuffer(make([]byte, 0, 2*sizeInMB*MiB))
+func (b *InMemoryWriteBuffer) WriteAt(content []byte, offset int64) error {
+	contentSize := int64(len(content))
+	if contentSize == 0 {
+		return nil
+	}
+
+	// Ensure b.currentBuffer != nil
+	if err := b.ensureCurrentBuffer(); err != nil {
+		return fmt.Errorf("ensureCurrentBuffer: %w", err)
+	}
+
+	if offset < b.current.offset.start ||
+		offset >= b.current.offset.end+b.bufferSize {
+		// TODO: finalise write and trigger temp file flow.
+		logger.Debugf("Non-sequential write encountered. TODO: Switch to " +
+			"temp-file flow insteading of erroring out.")
+		return fmt.Errorf(NonSequentialWriteError)
 	}
+
+	var contentWrittenSoFar int64
+	for {
+		n := b.writePartialContentToBuffer(content[contentWrittenSoFar:], offset+contentWrittenSoFar)
+		contentWrittenSoFar += n
+		if contentWrittenSoFar == int64(len(content)) {
+			// all content written successfully to buffer.
+			return nil
+		}
+		//TODO: Wait/timeout on last write to GCS. If unsuccessful, trigger temp file flow.
+		logger.Debugf("TODO: Wait/timeout on last write to GCS and if " +
+			"unsuccessful, trigger temp file flow.")
+		if err := b.advanceToNextChunk(); err != nil {
+			return fmt.Errorf("advanceToNextChunk: %w", err)
+		}
+		// TODO: trigger async upload to GCS for flushedBuffer.
+		logger.Debugf("TODO: trigger async upload to GCS for flushedBuffer.")
+	}
+}
+
+// Helper method to copy received content to currentBuffer.
+func (b *InMemoryWriteBuffer) copyDataToBuffer(contentStartOffset int64, content []byte) {
+	contentSize := int64(len(content))
+	contentEndOffset := contentStartOffset + contentSize
+	si := contentStartOffset % b.bufferSize
+	ei := si + contentSize
+
+	copy(b.current.buffer[si:ei], content)
+	b.updateFileSize(contentEndOffset)
 }
 
-func (b *InMemoryWriteBuffer) WriteAt(data []byte, offset int64) error {
+func (b *InMemoryWriteBuffer) advanceToNextChunk() error {
+	// ensure b.flushed.buffer != nil
+	if err := b.ensureFlushedBuffer(); err != nil {
+		return fmt.Errorf("ensureFlushedBuffer: %w", err)
+	}
+
+	// Move current buffer to flushed buffer.
+	b.flushed.buffer, b.current.buffer = b.current.buffer, b.flushed.buffer
+	b.flushed.offset = b.current.offset
+
+	// Make current buffer ready for new writes coming from kernel.
+	// TODO: revisit this when implementing on-disk-write-buffer.
```

Comment:
just to make it clear (no-pun intended)

```suggestion
	// TODO: revisit this decision to clear, when implementing on-disk-write-buffer.
```